crates/lang/codegen/src/generator/storage_item.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2018-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::GenerateCode;
+use derive_more::From;
+use ir::Selector;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{
+    format_ident,
+    quote,
+    ToTokens,
+};
+use syn::{
+    parse2,
+    GenericParam,
+};
+
+/// Generates code for the storage item.
+#[derive(From, Copy, Clone)]
+pub struct StorageItem<'a> {
+    /// The storage item to generate code for.
+    item: &'a ir::StorageItem,
+}
+
+impl GenerateCode for StorageItem<'_> {
+    /// Generates ink! storage item code.
+    fn generate_code(&self) -> TokenStream2 {
+        let generated_struct = self.generate_struct();
+        let generated_atomic_status = self.generate_atomic_status();
+        let generated_storage_type = self.generate_storage_type();
+        let generated_storage_key_holder = self.generate_storage_key_holder();
+
+        quote! {
+            #generated_struct
+            #generated_atomic_status
+            #generated_storage_type
+            #generated_storage_key_holder
+        }
+    }
+}
+
+impl<'a> StorageItem<'a> {
+    fn generate_struct(&self) -> TokenStream2 {
+        let item = self.item;
+        let ident = item.ident();
+        let attrs = item.attrs();
+        let vis = item.vis();
+        let generics = item.generics();
+        let salt = item.salt();
+
+        let fields = item.fields().map(|field| {
+            let key_bytes: Vec<u8> = if let Some(field_ident) = &field.ident {
+                [
+                    &ident.to_string().into_bytes()[..],
+                    &field_ident.to_string().into_bytes()[..],
+                ]
+                .concat()
+            } else {
+                ident.to_string().into_bytes()
```

Comment:
So tuple structs have colliding keys?

Reply:
In the case of tuple structs like `struct Test1(u32, u8)` yes, only the name of the struct defines the storage key, and if someone used two structs twice(or `Mapping`, or `StorageValue`), it generates the same storage key in each field.

So `struct Test1(Mapping<u32, u8>, Mapping<u128, u64>)` will work wrongly.

We can leave it as it is and emit errors during checking of storage key collisions, or add the restriction to not allow tuple structs.

Reply:
Another question how do we want to work with enums?
How should we calculate the storage key of each field in the variant?

It can be `NAME_OF_ENUM + NAME_OF_VARIANT + INDEX_OF_FIELD`.